Source/SteveSloth/Private/MyProjectileBaseClass.cpp

```diff
@@ -15,32 +15,213 @@ AMyProjectileBaseClass::AMyProjectileBaseClass()
 	{
+		Mesh = projectileData->Mesh;
+
 		ProjectileType = projectileData->ProjectileType;
 
 		Damage = projectileData->Damage;
 		ProjectileSpeed = projectileData->ProjectileSpeed;
+		ProjectileMaxSpeed = projectileData->ProjectileMaxSpeed;
 		ProjectileRange = projectileData->ProjectileRange;
 		ProjectileGravity = projectileData->ProjectileGravity;
 		InitialLaunchAngle = projectileData->InitialLaunchAngle;
 		AreaOfEffectDamage = projectileData->AreaOfEffectDamage;
 		AreaOfEffectRadius = projectileData->AreaOfEffectRadius;
 		DamageOverTime = projectileData->DamageOverTime;
+		DamageOverTimeRate = projectileData->DamageOverTimeRate;
 		DamageOverTimeDuration = projectileData->DamageOverTimeDuration;
 	}
+
+	Mesh = CreateDefaultSubobject<UStaticMeshComponent>(TEXT("Mesh"));
+
+	AreaOfEffectHitbox->SetSphereRadius(AreaOfEffectRadius);
+	AreaOfEffectHitbox->SetActive(false);
+
+	ProjectileMovement->bRotationFollowsVelocity = true;
+	ProjectileMovement->InitialSpeed = ProjectileSpeed;
+	ProjectileMovement->MaxSpeed = ProjectileMaxSpeed;
+	ProjectileMovement->ProjectileGravityScale = ProjectileGravity;
+	ProjectileMovement->Velocity.Z = InitialLaunchAngle;
 }
 
 void AMyProjectileBaseClass::BeginPlay()
 {
 	Super::BeginPlay();
+
+	Player = USteveSingleton::GetSteve()->GetPlayerCharacter();
+
+	if (IsValid(Player))
+	{
+		Steve = Cast<AMyPlayer>(Player);
+	}
+
+	StartingLocation = GetActorLocation();
+
+	ProjectileHitbox->OnComponentBeginOverlap.AddDynamic(this, &AMyProjectileBaseClass::OnHitboxOverlapBegin);
+	AreaOfEffectHitbox->OnComponentBeginOverlap.AddDynamic(this, &AMyProjectileBaseClass::OnAOEHitboxOverlapBegin);
 }
 
 void AMyProjectileBaseClass::Tick(float DeltaTime)
 {
 	Super::Tick(DeltaTime);
-}
\ No newline at end of file
+
+	if (FVector::Dist(StartingLocation, GetActorLocation()) > ProjectileRange)
+	{
+		this->Destroy();
+	}
+}
+
+void AMyProjectileBaseClass::OnHitboxOverlapBegin(UPrimitiveComponent* OverlappedComp, AActor* OtherActor, UPrimitiveComponent* OtherComp, int32 OtherBodyIndex, bool bFromSweep, const FHitResult& SweepResult)
+{
+	if (Enemy == OtherActor)
+	{
+		Enemy = Cast<AMyEnemyBaseClass>(OtherActor);
+
+		Enemy->HitEnemy(Damage);
+
+		switch (ProjectileType)
+		{
+		case EProjectileType::Physical:
+			this->Destroy();
+
+			break;
+
+		case EProjectileType::Fire:
+			Mesh->SetVisibility(false);
+			ProjectileHitbox->SetActive(false);
+			AreaOfEffectHitbox->SetActive(true);
+
+			break;
+
+		case EProjectileType::Water:
+			//add confusion here
+			this->Destroy();
+
+			break;
+
+		case EProjectileType::Poison:
+			Mesh->SetVisibility(false);
+			ProjectileHitbox->SetActive(false);
+
+			GetWorldTimerManager().SetTimer(DamageOverTimeTimerHandle, this, &AMyProjectileBaseClass::DamageOverTimeEnemy,
+				DamageOverTimeRate, true, DamageOverTimeRate);
+
+			break;
+
+		case EProjectileType::Ice:
+			Enemy->SetIsFrozen(true);
+			this->Destroy();
+
+			break;
+		}
+	}
+	else if (Steve == OtherActor)
```

Comment:
if (Steve == OtherActor) could we write theses the other way?
if (OtherActor == Steve) makes more sense to me this way..